### ISSUE TITLE
Train model using 3+ channels 

### ISSUE THREAD

**Issue by olgaliak** [matterport/Mask_RCNN#314]
Hello!
I'd like to use Mask RCNN for satellite imagery analysis.  In addition to  aerial view (RGB input), I want model to learn from hillshade view (thus learn topography) as well as Indexes data (NWDI -- Normalized Difference Water Index, EVI --Enhanced Vegetation Index).
Could you please provide guidance how to update the code to support N channels (instead of 3)?
Also ideally I'd love  to use pretrained weights when possible rather start completely from scratch all the time.
Thanks a lot!

**Comment by FruVirus** [comment]
@olgaliak 

Here are at least two areas that you have to change:

1) In config.py, you have to add extra dimensions to MEAN_PIXEL corresponding to the mean pixel value for each new channel you add (and make sure they are in the correct ordering). This is assuming you will be subtracting the mean pixel value per channel for your training.

2) In config.py, change the last dimension of self.IMAGE_SHAPE from 3 to N.

Assuming that your data pre-processing is ok, then in theory, the ResNet should build itself accordingly and the RPN, FPN, and ODN should follow suit.

**Comment by YubinXie** [comment]
I have 7 channels images and I successfully made 7-channel training work. It gives reasonable results but not as good as RGB.

**Comment by CMCDragonkai** [comment]
The `utils.py` `load_image` function also currently removes the alpha channel if it is the last channel. So you also have to modify the `load_image` function.

**Comment by qmzsuxing** [comment]
@YubinXie how did you do that？could you please tell me more about that? my email: 1003810718@qq.com

**Comment by qmzsuxing** [comment]
@FruVirus  If I use the Facebook/Detectron, how can I change the code to support N channels (instead of 3)?

**Comment by YubinXie** [comment]
Just change the code that about dimension in model.py (first layer of ResNet) and config.py accordingly.

**Comment by qmzsuxing** [comment]
@YubinXie could you please give me your email address?

**Comment by qmzsuxing** [comment]
@YubinXie and my wechat: qinmengzhe

**Comment by qmzsuxing** [comment]
@waleedka Thank you so much!

**Comment by YubinXie** [comment]
@qmzsuxing You can post your questions in this repo or create a new issue. Others and I will try our best to help you (though I am not an expert in this area). By doing this, your questions and the answers that you get will also help other users.

**Comment by qmzsuxing** [comment]
@olgaliak Hi, I wonder when training satellite images, which backbone did you choose? resnet50 or resnet101？ When I train the satellite images upon resnet101, I always get a large val_loss...

**Comment by Xiaoyang-Rebecca** [comment]
> @olgaliak Hi, I wonder when training satellite images, which backbone did you choose? resnet50 or resnet101？ When I train the satellite images upon resnet101, I always get a large val_loss...

resnet50 is a lighter backbone than resnet101, it should converge faster than the latter.

**Comment by orestis-z** [comment]
Hi, if you need a specific example, you can have look at this code snippet https://github.com/orestis-z/mask-rcnn-rgbd/blob/d590e0f5085f8cbe895a6698e284426fd0116aa4/instance_segmentation/sceneNet/train.py#L66-L85 for how to extend the first convolution for RGB-D data by setting the weights of the depth channel to the mean of the RGB channels
Cheers
